### PR TITLE
[23.1] Skip change_datatype things if we're not actually changing the extension

### PR DESCRIPTION
**Pull request by mvdbeek**
Side-steps a problem with FileParameter in the most efficient way possible. This likely became a problem for one of Wolfgang's workflow after we dropped some earlier unnecessary flushes.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/model/metadata.py [hunk 1]
```diff
@@ -604,7 +604,16 @@ def wrap(self, value, session):
         if isinstance(value, int):
             return session.query(galaxy.model.MetadataFile).get(value)
         else:
-            return session.query(galaxy.model.MetadataFile).filter_by(uuid=value).one()
+            wrapped_value = session.query(galaxy.model.MetadataFile).filter_by(uuid=value).one_or_none()
```
Comment (jdavcs) [inline lib/galaxy/model/metadata.py:607]: Maybe factor out into a function? Then the logic is so much easier to read, I think:
```py
wrapped_value = get_wrapped_value(session, value)
if wrapped_value:
    return wrapped_value
else:
    # commit here
    return get_wrapped_value(session, value)
...
def get_wrapped_value(session, value):
    stmt = select(galaxy.model.MetadataFile).filter_by(uuid=value)
    return session.scalars(stmt).one_or_none()
```

**Review by jdavcs — APPROVED** [review]

**Review by martenson — APPROVED** [review]

**Comment by jdavcs** [comment]
@martenson @mvdbeek there was an unresolved comment.

File: lib/galaxy/model/metadata.py (hunk 1 above)
Comment (martenson) [reply lib/galaxy/model/metadata.py:607]: This is already conflicting (https://github.com/galaxyproject/galaxy/pull/16933) bugfix on a released branch on notoriously problematic module. I think refactoring should go against dev afterwards. That said, I am sorry I did not communicate this clearly before merging.

I will open that refactoring PR once we resolve the conflict.

Comment (jdavcs) [reply lib/galaxy/model/metadata.py:607]: No problem, I can fix this in a follow-up. It was not refactoring - it was new code added (minus sqlalchemy 2.0 syntax), which is why I suggested the change. Again, that's OK. I'll take a look at the conflict.

**Pull request merged**